### ISSUE TITLE
AttributeError: module 'pyscf.tblis_einsum' has no attribute 'tblis_dtype'

### ISSUE THREAD

**Issue by vvp-nsk** [BoothGroup/ebcc#41]
Hej!

Please correct me if I am wrong, but EBCC might run faster if it utilizes the TBLIS library via PySCF's wrapper to it, right?  I just locally installed TBLIS and pyscf-tblis  but EBCC won’t work:

```
ccsd_sp.kernel()
  File "/home/rad/ebcc/ebcc/rebcc.py", line 364, in kernel
    e_cc = e_init = self.energy(amplitudes=amplitudes, eris=eris)
  File "/home/rad/ebcc/ebcc/rebcc.py", line 776, in energy
    return func(**kwargs)
  File "/home/rad/ebcc/ebcc/codegen/UCC2.py", line 10, in energy
    e_cc += einsum(t2.aaaa, (0, 1, 2, 3), v.aaaa.ovov, (0, 3, 1, 2), ()) * -1.0
  File "/home/rad/ebcc/ebcc/util.py", line 1108, in einsum
    out = _contract(subscript, *args, **kwargs)
  File "/home/rad/ebcc/ebcc/util.py", line 1041, in contract
    tblis_dtype = tblis_einsum.tblis_dtype[dtype]
AttributeError: module 'pyscf.tblis_einsum' has no attribute 'tblis_dtype
```

Do you know by chance what could be wrong with my local installation of TBLIS? 

Thank you in advance!

With best regards, 
Victor

**Comment by obackhouse** [comment]
Should be fixed in #42, I'll merge once checks pass. Thanks, this is my oversight as my TBLIS is configured separately from PySCF. The `ebcc.util` code is a complete mess that I need to get round to cleaning up...

Indeed, I think that TBLIS has potential to remove speeds. I would expect this to be most dramatic in the lower scaling ansatzes (particular CC2 like you appear to be using) because the memory copying for non-TBLIS `einsum` implementations going via DGEMM will never be worse than N^4. Let me know how it goes.

**Comment by vvp-nsk** [comment]
Hej!

Thanks for the prompt fix. Now everything works out. The CC2 performs quite impressive, IMHO. 
I have a naive question. Does EBCC support the B-UCC2 (spin-unrestricted [Brueckner CC2](https://www.sciencedirect.com/science/article/pii/S0009261411002193)) ansatz?  

With best regards, 
Victor

**Comment by vvp-nsk** [comment]
Hej!

I slightly modified your example to generate a real UHF open-shell reference case:

```
mport numpy as np
from pyscf import gto, scf

from ebcc import UEBCC

# Define the molecule using PySCF
mol = gto.Mole()
mol.atom = "H 0 0 0; F 0 0 1.1"
mol.basis = "cc-pvdz"
mol.charge = 1
mol.spin = 1
mol.build()

# Run a UHF calculation using PySCF
mf = scf.UHF(mol)
#mf = mf.density_fit()
mf.kernel()

# Run a CCSD calculation
ccsd = UEBCC(mf, ansatz="CC2")
ccsd.kernel()

# Run a Brueckner orbital calculation using the CCSD reference
ccsd.brueckner(e_tol=1e-6, t_tol=1e-5)
```

The input given above fails with the following error:
```
ebcc/ebcc/brueckner.py", line 410, in get_rotation_matrix
    a = diis.update(a, xerr=np.array([t1.aa, t1.bb]))
ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (2,) + inhomogeneous part.
```

Any help will be greatly appreciated!

With best regards, 
Victor

**Comment by obackhouse** [comment]
> Hej!
> 
> I slightly modified your example to generate a real UHF open-shell reference case:
> 
> ```
> mport numpy as np
> from pyscf import gto, scf
> 
> from ebcc import UEBCC
> 
> # Define the molecule using PySCF
> mol = gto.Mole()
> mol.atom = "H 0 0 0; F 0 0 1.1"
> mol.basis = "cc-pvdz"
> mol.charge = 1
> mol.spin = 1
> mol.build()
> 
> # Run a UHF calculation using PySCF
> mf = scf.UHF(mol)
> #mf = mf.density_fit()
> mf.kernel()
> 
> # Run a CCSD calculation
> ccsd = UEBCC(mf, ansatz="CC2")
> ccsd.kernel()
> 
> # Run a Brueckner orbital calculation using the CCSD reference
> ccsd.brueckner(e_tol=1e-6, t_tol=1e-5)
> ```
> 
> The input given above fails with the following error:
> 
> ```
> ebcc/ebcc/brueckner.py", line 410, in get_rotation_matrix
>     a = diis.update(a, xerr=np.array([t1.aa, t1.bb]))
> ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (2,) + inhomogeneous part.
> ```
> 
> Any help will be greatly appreciated!
> 
> With best regards, Victor

Ahh I see the issue, I've coded it up badly. This will take me a little longer, I'll try to get a fix later today. Sorry about all the bugs, I appreciate you identifying them though!

**Comment by obackhouse** [comment]
Should be fixed in #45. If you have any more problems, keep them coming :smile:
